packages/embark-ui/src/services/web3.js

```diff
@@ -21,7 +21,9 @@ export function estimateGas({web3, contract, args}) {
 
 export function deploy({web3, contract, args}) {
   return new Promise((resolve, reject) => {
-    const cleanup = () => { promiEvent.removeAllListeners(); };
+    const cleanup = async () => { 
+      (await promiEvent).removeAllListeners(); 
```

Comment:
Cancelling a contract deployment (Deployment tab in Cockpit using injected web3), caused an error with `removeAllListeners` as `promiEvent` is a promise, and it was not resolved yet, causing `removeAllListeners` to be `not a function`.

Reply:
When you say injected web3, is that metamask's web3?  The API changes w/ 1.0.0 series; prior to that the promiEvent hybrid object didn't exist?

Reply:
@michaelsbradleyjr - yes, injected web3 is using metamask or mist. 

I've updated this so that the `promiEvent` is `Promise.resolve`'d regardless, and then the resolved value has it's listeners removed. This should work for pre-1.0, correct?

Reply:
1.0 docs for the [return value of the `send`](https://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#id22), which is a method of the transaction object returned by the `deploy` method.

Looking at the docs for pre-1.0, it's not clear to me how it can work because [`web3.eth.Contract`](https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethcontract) instances don't seem to have a `deploy` method, but there may be a misunderstanding on my part. When I [revised the code related to the cleanup function](https://github.com/embark-framework/embark/pull/1323/files#diff-ad35638d566c3c0c3320d0ea77661196R22), I recall looking at the older docs and thinking that we must have already been assuming the web3 version is recent enough to have the expected methods.

Reply:
Also, for the 1.0 case, I don't think it's correct to await the Promise before removing all listeners.

Reply:
In the context of 1.0+ web3, what do you suggest is the best way? If web3 <1.0 does not have a `.deploy` method, then it looks like we are currently not supporting that case.

Reply:
For 1.0+, I think it can be how it was before:
```js
const cleanup = () => { promiEvent.removeAllListeners(); };
```
The reason is that [cleanup is invoked when promiEvent resolves/rejects](https://github.com/embark-framework/embark/blob/0223dd9b8317f64fab42f1fcfbe9f87babd2e4a0/packages/embark-ui/src/services/web3.js#L32-L33), so no need to `await` it again.

Reply:
Sorry, going back to your original comment and giving it more thought, to make sure that matter is addressed.

Reply:
Was the problem mentioned in your original comment encountered with web3 1.0+ or only when it's pre-1.0?

Reply:
> For 1.0+, I think it can be how it was before:
```js
const cleanup = () => { promiEvent.removeAllListeners(); };
```

> The reason is that [cleanup is invoked when promiEvent resolves/rejects](https://github.com/embark-framework/embark/blob/0223dd9b8317f64fab42f1fcfbe9f87babd2e4a0/packages/embark-ui/src/services/web3.js#L32-L33), so no need to `await` it again.

This is erroring with 1.0, hence the need for this PR. It has been replicated by a few others as well.

Reply:
What if we make the following changes?

```
const noop = () => {};
const promiEvent = new web3.eth.Contract(contract.abiDefinition)
  .deploy({data: `0x${contract.code}`, arguments: args})
  .send({from: web3.eth.defaultAccount})
  .once('error', reject)
  .once('receipt', resolve)
  .then(noop)
  .catch(noop);
```

The outer promise returned by `deploy` completely encapsulates `promiEvent`, i.e. its methods and events are not accessible to the caller. That was the case even prior to #1323.

My idea with `cleanup` was to ensure `promiEvent` gets gc'd.  Maybe that can be accomplished by switching from `on` to `once` for the `'error'` and `'receipt'` events, which reject/resolve the outer promise. And we avoid unhandled rejection of the inner promise by `.catch(noop)`. Resolution of the inner promise is handled by `.then(noop)` for good measure.

Reply:
I suspect that `.then(cleanup)` was happening prior to `.on('receipt', resolve)`.

Reply:
The 1.0 docs say re: the `error` event:
> "error" returns Error: is fired if an error occurs during sending. If an out of gas error, the second parameter is the receipt.

The outer promise doesn't currently handle a case when there are two arguments passed to the callback of the `'error'` event, which is the outer promise's `reject` function.

To fully handle that, we'd need to replace the `'error'` callback with something like:

```js
function (...args) {
  args.length < 2 ? reject(args[0]) : reject(args);
}
```

Also, the caller of the exported `deploy` would need to test with `Array.isArray` in its rejection handler.

Reply:
This has been updated, choosing to put the `noop` function directly in `.then()`, and `reject` in the `.catch`.

Reply:
💡💡💡 I think I just realized why `cleanup` was failing. Consider the original code:

```js
const cleanup = () => { promiEvent.removeAllListeners(); };
const promiEvent = new web3.eth.Contract(contract.abiDefinition)
  .deploy({data: `0x${contract.code}`, arguments: args})
  .send({from: web3.eth.defaultAccount})
  .on('error', reject)
  .on('receipt', resolve)
  .then(cleanup)
  .catch(cleanup);
```

I think what's happening is that `.then` doesn't return the combo instance. Rather, it's a method baked into the promise (into which [the EventEmitter methods have been smushed](https://github.com/ethereum/web3.js/blob/1.0.0-beta.37/packages/web3-core-promievent/src/index.js#L35-L60)).

So, when calling `.then` and `.catch` in the method chain, `const promiEvent` ends being assigned a plain promise, not the combo.

Here's a revision that I think will work:

```js
const cleanup = () => { promiEvent.removeAllListeners(); };
const promiEvent = new web3.eth.Contract(contract.abiDefinition)
  .deploy({data: `0x${contract.code}`, arguments: args})
  .send({from: web3.eth.defaultAccount})
  .on('error', reject)
  .on('receipt', resolve)

promiEvent
  .then(cleanup)
  .catch(cleanup);
```

That way `promiEvent` is the combo instance, we're guaranteed to have cleaned up all the event handlers, and we've covered all the bases of inner promise resolution and rejection.

Reply:
After trying this, it doesn't quite work. Additions/updates needed:
1. `error` needs to be fired more than once. This is due to any number of errors that we may get from the blockchain before a tx is confirmed.
2. `cleanup` doesn't seem to be needed. Looking [in the source at `.deploy`](https://github.com/ethereum/web3.js/blob/1d9f6c0889c7f551b4ec1041cece0d50bc3ff2c8/packages/web3-eth-contract/src/index.js#L479), it says `All event listeners will be removed, once the last possible event is fired ("error", or "receipt")`, which seems like it does the work for us (although honestly I had difficult time finding the code that actually does this). Therefore, leaving the majority of our work in `.on('error')` and `.once('receipt')`, and handling any stranglers with `.catch(reject)` and `.then(() => {})` should be more than sufficient.

In the latest (tested) commit, the deploy function looks like:

```
export function deploy({web3, contract, args}) {
  return new Promise((resolve, reject) => {
    new web3.eth.Contract(contract.abiDefinition)
      .deploy({data: `0x${contract.code}`, arguments: args})
      .send({from: web3.eth.defaultAccount})
      .on('error', reject)
      .once('receipt', resolve)
      .catch(reject)
      .then(() => {});
    });
}
```

I think we have spent quite a bit of time on this already, and after some thorough testing, we haven't really seen any nasties, but I'd encourage you to test this yourself.

1. Run the demo, go to Cockpit > Deployment
2. Ensure Metamask is enabled. 
3. Select the option "Injected Web3". If you have private mode enabled on Metamask, you should see a popup asking if it's OK to connect the website to Metamask. Click "Connect" if you see this.
4. Type a number value for the contract constructor, then click the "Deploy" button.
5. For the first test, click "Reject" in the resulting Metamask popup. You should see the error appear `Deployment failed: Returned error: Error: MetaMask Tx Signature: User denied transaction signature.` in the cockpit.
6. For the second test, click "Deploy" again, then in the resulting Metamask popup, click "Confirm".  In the cockpit, you should see "Deployment succeed:" with details of the deployment. If you having trouble getting the deployment to succeed, you may need to enable regular transactions by typing `regularTxs on` in the console.

Reply:
Sorry, I didn't see your reply comment until just now.

> Looking in the source at .deploy, it says All event listeners will be removed, once the last possible event is fired ("error", or "receipt"), which seems like it does the work for us...

Excellent!  My apologies for missing that earlier.  Indeed, if that's the case then we don't need to worry about gc: `cleanup` isn't necessary, `.catch(reject)` and `.then(() => {})` are the best way to go.  Sorry for dragging this out, I was just worried that in long-running sessions of cockpit, we could have a leak if we didn't clean up the event handlers manually.

> error needs to be fired more than once. This is due to any number of errors that we may get from the blockchain before a tx is confirmed.

That's actually impossible with respect to the outer promise, since once the outer promise is rejected, it's "locked". Even if `reject` is fired as the callback of the `'error'` event multiple times, the outer promise will only reflect the first call of `reject`.  The same is true of `resolve`.

If potentially multiple errors may fire on `promiEvent` and if it's important we handle/report all of them, then instead of a promise the exported `deploy` could return an rxjs observable (or an event emitter); alternatively, exported `deploy` could take a callback that is understood to possibly be invoked multiples times. Yet another alternative is to simply return the combined Promise-EventEmitter instance (`promiEvent`) and leave it to the caller to handle the difference between the resolved value and the object that comes through to a callback of the `receipt` event. In other words, our wrapping with an outer promise may be a poor abstraction for what we're actually dealing with.

To see the behavior of reject/resolve described above, try this in a REPL:

```js
let reject;

let p = new Promise((res, rej) => {
  reject = rej;
});
p.then(v => console.log(v)).catch(e => console.error(e.message));

reject(new Error('1st error'));  // prints "1st error"
reject(new Error('2nd error'));  // nothing happens
reject(new Error('3rd error'));  // nothing happens

let resolve;

p = new Promise((res, rej) => {
  resolve = res;
});
p.then(v => console.log(v)).catch(e => console.error(e.message));

resolve('1st value');  // prints "1st value"
resolve('2nd value');  // nothing happens
resolve('3rd value');  // nothing happens
```

Reply:
For clarification of users who may be reading this later, there will only ever be one error event fired, which is hopefully clarified above.

After discussing offline, @michaelsbradleyjr agrees this is OK to merge.